TESTS/mbed_hal/uart_bare_metal_test/main.cpp

```diff
@@ -0,0 +1,152 @@
+
+/* UART TEST FOR GREEN TEA
+ *
+ * Greentea-client uses RawSerial class to handle UART communication.
+ * RawSerial class must provide valid implementation of the following
+ * methods (used by Greentea-client):
+ *  - int RawSerial::getc()
+ *  - int RawSerial::putc(int c).
+ *
+ * Greentea-client uses USB serial port for communication with the
+ * following transmission parameters:
+ *  - 8 data bits,
+ *  - 1 stop bit,
+ *  - none parity bit,
+ *  - configurable baudrate.
+ *
+ * Greentea-client operates in pooling mode only.
+ * Flow control is disabled.
+ *
+ * UART buffer size is limited to 120 bytes.
+ *
+ * TEST SCENARIO
+ * This part of the test is executed on mbed board.
+ *
+ * After startup it will try to communicate via UART with host part of the test.
+ * This part sends 4 messages with different length to the host and waits
+ * for the response (host should respond with the same message).
+ * Last message "PASSED" is send to inform that all responses from host were
+ * successfully received.
+ *
+ * To perform this test with serial port baud rate different than default (9600),
+ * call host test with appropriate --baud rate parameter.
+ *
+ */
+
+#define NUMBER_OF_TEST_STRINGS 5
+#define NUMBER_OF_TEST_RETRIES 10
+
+typedef enum
+{
+    PASSED, FAILED
+} test_status;
+
+/* Array of string to be transmitted. */
+const char *test_strings[NUMBER_OF_TEST_STRINGS] =
+        { "x", "mbed", "Green Tea UART bear metal test.",
+            "Long Message-Long Message-Long Message-Long Message-Long Message-Long Message-Long Message-Long Message-Long Message--.",
+            "PASSED" };
+
+/* Function writes provided string to the serial port. */
+void write_line(RawSerial &serial, const char *str)
+{
+    while (true) {
+        serial.putc(*str);
+        if (*str == '\0') {
+            break;
+        }
+
+        str++;
+    };
+}
+
+/* Function reads provided string from the serial port. */
+int read_line(RawSerial &serial, char *buffer)
+{
+    int count = 0;
+    while (true) {
+        *buffer = (char) serial.getc();
+        count++;
+        if (*buffer == '\0') {
+            return count;
+        }
+
+        buffer += 1;
+    }
+}
+
+/* Function compares provided strings and returns 0 if are equal. */
+int strCmp(const char string1[], char string2[])
+{
+    for (int i = 0;; i++) {
+        if (string1[i] != string2[i]) {
+            return string1[i] < string2[i] ? -1 : 1;
+        }
+
+        if (string1[i] == '\0') {
+            return 0;
+        }
+    }
+}
+
+/* Function writes provided message to the serial port and
+ * waits for the response from the Host. Host should
+ * respond with the same message.
+ *
+ * If valid response has been received, then function returns
+ * PASSED status, otherwise transmission is repeated.
+ *
+ * If max number of transmission attempts is reached, then
+ * function returns FAILED status.
+ *
+ * */
+test_status test_uart_communication(RawSerial &serial, const char * msg)
```

Comment:
doxygen comments, in separate header file?

Reply:
I'm not sure if we need this for this test. This test must be executed manually and does not use grean tea. The hart of this test is the python script and it provides detailed description.